### PR TITLE
Add an FST index to `ImportMap` and use it to speed up auto import

### PR DESCRIPTION
**Pull request by jonas-schievink**
For the importing crate, we still use the symbol index, but I've modified it to only look at files that comprise that crate (instead of the whole workspace).

Oh, and since now the symbol query limit is respected correctly, it's possible that some results from the local crate now disappear if there are many matches.

Fixes https://github.com/rust-analyzer/rust-analyzer/issues/4763

### CODE REVIEW

File: crates/ra_ide_db/src/symbol_index.rs [hunk 1]
```diff
@@ -173,6 +176,30 @@ pub fn world_symbols(db: &RootDatabase, query: Query) -> Vec<FileSymbol> {
...
 
+pub fn crate_symbols(db: &RootDatabase, krate: CrateId, query: Query) -> Vec<FileSymbol> {
+    let def_map = db.crate_def_map(krate);
+    let mut files = Vec::new();
+    let mut modules = vec![def_map.root];
+    while let Some(module) = modules.pop() {
+        let data = &def_map[module];
+        files.extend(data.origin.file_id());
+        modules.extend(data.children.values());
+    }
```
Comment (matklad) [inline crates/ra_ide_db/src/symbol_index.rs:187]: Hm, if we are looking at `def_map` anyway, maybe we should just dump the symbols from it directly (as opposed to collecting the files, and *then* getting symbols from files' syntax)?

File: crates/ra_ide_db/src/symbol_index.rs [hunk 2]
```diff
@@ -312,6 +336,10 @@ impl Query {
                     if self.exact && symbol.name != self.query {
                         continue;
                     }
+
+                    if res.len() >= self.limit {
+                        return res;
+                    }
                     res.push(symbol.clone());
```
Comment (matklad) [inline crates/ra_ide_db/src/symbol_index.rs:343]: Seems like this should be moved *before* the `if`?

File: crates/ra_hir_def/src/import_map.rs [hunk 3]
```diff
@@ -117,13 +163,97 @@ impl fmt::Debug for ImportMap {
...
+/// Searches dependencies of `krate` for an importable path matching `query`.
+///
+/// This returns all items that could be imported from within `krate`, excluding paths inside
+/// `krate` itself.
+pub fn search_dependencies<'a>(
+    db: &'a dyn DefDatabase,
+    krate: CrateId,
+    query: Query,
+) -> Vec<ItemInNs> {
+    let _p = ra_prof::profile("import_map::global_search").detail(|| format!("{:?}", query));
```
[the remark this answers is not in the document]
Comment (jonas-schievink) [reply crates/ra_hir_def/src/import_map.rs:207]: Oh yeah, I missed that while renaming the function

File: crates/ra_ide_db/src/imports_locator.rs [hunk 4]
```diff
@@ -11,44 +11,46 @@ use crate::{
...
-        let project_results = {
-            let mut query = Query::new(name_to_import.to_string());
-            query.exact();
-            query.limit(40);
-            symbol_index::world_symbols(db, query)
-        };
-        let lib_results = {
+        // Query dependencies first.
+        let mut candidates: FxHashSet<_> =
+            self.krate.query_external_importables(db, name_to_import).collect();
```
[the remark this answers is not in the document]
Comment (jonas-schievink) [reply crates/ra_ide_db/src/imports_locator.rs:32]: I put in a query limit, which is set to 40, like it was before

File: crates/ra_hir/src/code_model.rs [hunk 5]
```diff
@@ -98,6 +99,23 @@ impl Crate {
         db.crate_graph()[self.id].display_name.as_ref().cloned()
     }
 
+    pub fn query_external_importables(
+        self,
+        db: &dyn DefDatabase,
+        query: &str,
+    ) -> impl Iterator<Item = Either<ModuleDef, MacroDef>> {
```
Comment (matklad) [inline crates/ra_hir/src/code_model.rs:106]: Note: we might or might not want to refactor `ModuleDef` to to contain `MacroDef` (but keeeping ModuleDefId macro-free).

Comment (matklad) [reply crates/ra_hir/src/code_model.rs:106]: Or maybe even go further and remove distinction between `ScopeDef` and `ModuleDef` altogether? I feel for IDE cases it maybe makes sense to not do fine-grained distinction here? We already have `Definition` in `ra_ide_db` for that purpose, but I am wondering if we should push it down to the hir layer?

File: crates/ra_ide_db/src/symbol_index.rs (hunk 1 above)
Comment (matklad) [reply crates/ra_ide_db/src/symbol_index.rs:187]: @jonas-schievink are we leaving this for the future work?

Comment (jonas-schievink) [reply crates/ra_ide_db/src/symbol_index.rs:187]: I think so, yes. I can put in a FIXME or open an issue.

Comment (matklad) [reply crates/ra_ide_db/src/symbol_index.rs:187]: r=me with an issue linked!

**Comment by jonas-schievink** [comment]
bors r=matklad

**Pull request merged**